ArcGISRuntimeSDKQt_CppSamples/Analysis/StatisticalQueryGroupSort/StatisticalQueryGroupSort.cpp

```diff
@@ -97,26 +100,24 @@ void StatisticalQueryGroupSort::connectSignals()
     while (iter.hasNext())
     {
       // get the statistic record
-      StatisticRecord* record = iter.next(result);
+      StatisticRecord* record = iter.next(result.data());
 
       // get the group string
       QString sectionString;
-      for (auto group : record->group().keys())
+      for (const QString& group : record->group().keys())
       {
         if (sectionString.length() > 0)
           sectionString += ",";
         sectionString += QString("\"%1\":\"%2\"").arg(group, record->group().value(group).toString());
       }
 
       // obtain the statistics
-      for (auto stats : record->statistics().keys())
+      for (const QString& stats : record->statistics().keys())
```

Comment:
This is optional and you could argue it makes the code harder to follow, but `collection.keys()` is one of the clazy warnings that will come up with a code scan.

See [here](https://github.com/KDE/clazy/blob/master/docs/checks/README-container-anti-pattern.md).

You can get around this by using iterators. However, although it's more efficient I think it reads nicer like this.

Reply:
If I had to choose here, I'd leave the code as its written. Efficiency is less important than readable code here.

Reply:
how about:

```
const QStringList recordKeys = record->statistics().keys();
for (const QString& stats : recordKeys )
```

that seems more readable and gets around the warning I think?

Reply:
That would have the same warning. It's the `.keys()` that's the culprit. The issue is that the keys are already stored in record->statistics map, and calling .keys on that allocates a new list when you already have that info.

It would have to be something like this.
```cpp
const QVariantMap statsMap = record->statistics();
for (auto it = statsMap.cbegin(); it != statsMap.cend(); ++it)
{
  const QString statString = QString("%1: %2").arg(it.key(), it.value().toString());
  addResultToModel(sectionString, statString);
}
```

I don't know if that's more readable or not. The most important thing for the samples is readability.

Reply:
@JamesMBallard  I think that is easier to read. The mental load for what the format arguments are doing is much easier to understand. Unless there are objects I'll probably use this suggestion here.

Reply:
Sounds good to me @anmacdonald. I think there are a few places in this file that use the same pattern.

Reply:
Just realized that now we are purging auto it would now read as...

```
for (QVariantMap::const_iterator it = statsMap.begin(); it != statsMap.end(); ++it)
{
....
```

Which is a bit more ungainly but if there are no objection that's how I'll format the for loop.

Reply:
keep the iterator as `auto`. It's not a runtime Api type, and this is where `auto` really shines. It's obvious it's an iterator based on the usage.

Reply:
Also, it would be good to use the const iterators, so `.cbegin()` and `.cend()`, respectively.